examples/jsm/renderers/nodes/core/NodeBuilder.js

```diff
@@ -0,0 +1,166 @@
+const VERSION = '1';
+
+class NodeBuilder {
+
+	constructor() {
+		
+		this.slots = { vertex: [], fragment: [] };
+		this.defines = { vertex: {}, fragment: {} };
+		
+		this.nodesData = new WeakMap();
+		
+		this.shader = undefined;
+		
+	}
+	
+	setMaterial( material ) {
+		
+		this.material = material;
+		
+	}
+	
+	addSlot( shader, slot ) {
+		
+		this.slots[ shader ].push( slot );
+		
+	}
+	
+	addDefine( shader, name, value ) {
+		
+		this.defines[ shader ][ name ] = value || '';
+		
+	}
+	
+	generateVec3( x, y, z ) {
+		
+		return `vec3( ${x}, ${y}, ${z} )`;
+		
+	}
+	
+	generateFloat( value ) {
+		
+		return value + ( value % 1 ? '' : '.0' );
+		
+	}
+	/*
+	createDataFromNode( node ) {
+		
+		return this.nodesData[ node ] = this.nodesData[ node ] || {};
+		
+	}
+	*/
+	createUniformFromNode( node ) {
+		
+		
+		
+	}
+	/*
+	analyzeNode( node ) {
+		
+		
+	}
+	*/
+	flowNode( node, output ) {
+		
+		let flowData = {};
+		flowData.result = node.build( this, output );
+		
+		return flowData;
+		
+	}
+	
+	buildDefines( shader ) {
+		
+		const defines = this.defines[ shader ];
+		
+		let code = '';
+		
+		for ( let name in defines ) {
+			
+			code += `#define NODE_${name} ${defines[name]}\n`;
+			
+		}
+		
+		return code;
+		
+	}
+	
+	build( shader ) {
+		
+		const slots = this.slots[ shader ];
+		
+		if ( slots.length ) {
+			
+			this.addDefine( shader, 'NODE', VERSION );
+			
+			for( let i = 0; i < slots.length; i++) {
+			
+				let slot = slots[i];
+				
+				let flowData = this.flowNode( slot.node );
+				
+				this.addDefine( shader, slot.name, flowData.result );
+				
+			}
+			
+		}
+		
+		let defines = this.buildDefines( shader );
+		
+		return {
+			defines
+		};
+		
+	}
+	
+	format( code, fromType, toType ) {
+
+		const typeToType = toType + ' <- ' + fromType;
+
+		switch ( typeToType ) {
+
+			case 'f <- v2' : return code + '.x';
```

Comment:
This code section is probably easier to read if you following this pattern: `from -> to`. So:
```js
const conversion = fromType + ' -> ' + toType;

switch ( conversion ) {

    case 'v2 -> f' : return code + '.x';
```